File: LCIntelliJ/src/LC90.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class LC90 {

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> list = new ArrayList<>();
        dfs(res, list, nums, 0);
        return res;
    }

    private void dfs(List<List<Integer>> res, List<Integer> list, int[] nums, int idx) {
        res.add(new ArrayList<Integer>(list));
        for(int i = idx; i < nums.length; i++) {
            if(i > idx && nums[i] == nums[i - 1]) {
                continue;
            }
            list.add(nums[i]);
            dfs(res, list, nums, i + 1);
            list.remove(list.size() - 1);
        }
    }


    public static void main(String[] args) {
        int[] nums = new int[] {2, 2, 1};
        LC90 obj = new LC90();
        List<List<Integer>> list = obj.subsetsWithDup(nums);
        PrintUtils.printListOfListInteger(list);
    }
}

File: facebook/OC/LC/LC/LC90.m
//
//  LC90.m
//  LC
//
//  Created by Zhen Gong on 5/1/18.
//  Copyright © 2018 ULS. All rights reserved.
//

#import "LC90.h"
#import "QuickSort.h"

@implementation LC90

- (NSArray *)subsetsWithDup:(int*)nums numsSize:(int)numsSize {
    NSMutableArray *res = [NSMutableArray array];
    NSMutableArray *list = [NSMutableArray array];
    [QuickSort quickSortInt:nums size:numsSize];
    [QuickSort printIntArray:nums size:numsSize];
    [self dfs:res list:list nums:nums numsSize:numsSize idx:0];
    return res;
}

- (void)dfs:(NSMutableArray *)res list:(NSMutableArray *)list nums:(int *)nums numsSize:(int)numsSize idx:(int)idx {
    [res addObject:[NSArray arrayWithArray:list]];
    [list enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        NSNumber *num = (NSNumber *)obj;
        DLog(@"%d", [num intValue]);
    }];
    
    for(int i = idx; i < numsSize; i++) {
        if(i > idx && nums[i - 1] == nums[i]) {
            continue;
        }
        [list addObject:@(nums[i])];
        [self dfs:res list:list nums:nums numsSize:numsSize idx:i + 1];
        [list removeLastObject];
    }
}

- (void)test {
    int array[] = {2, 2, 1};
    int* ptr = array;
    NSArray* res = [self subsetsWithDup:ptr numsSize:3];
    [res enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        NSArray *list = (NSArray *)obj;
        printf("[");
        [list enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            NSNumber *num = (NSNumber *)obj;
            printf("%d, ", [num intValue]);
        }];
        printf("]");
    }];
    
}

@end
